File: scripts/generate_quran_db.js
const sqlite3 = require('sqlite3').verbose();
const https = require('https');
const fs = require('fs');
const path = require('path');

function fetchJson(url) {
    return new Promise((resolve, reject) => {
        https.get(url, (res) => {
            if (res.statusCode !== 200) {
                reject(new Error(`HTTP Error: ${res.statusCode} for ${url}`));
                return;
            }
            let rawData = '';
            res.on('data', (chunk) => rawData += chunk);
            res.on('end', () => {
                try {
                    resolve(JSON.parse(rawData));
                } catch (e) {
                    reject(e);
                }
            });
        }).on('error', reject);
    });
}

async function main() {
    console.log("🚀 Starting Quran Database Generation...");

    const arabicUrl = "https://api.alquran.cloud/v1/quran/quran-uthmani";
    const persianUrl = "https://api.alquran.cloud/v1/quran/fa.ansarian";

    console.log(`📥 Fetching Arabic Uthmani payload...`);
    const arabicPayload = await fetchJson(arabicUrl);
    
    console.log(`📥 Fetching Persian Ansarian Translation payload...`);
    const persianPayload = await fetchJson(persianUrl);

    if (arabicPayload.code !== 200 || !arabicPayload.data || !arabicPayload.data.surahs) {
        console.error("❌ Invalid Arabic API payload!");
        process.exit(1);
    }

    if (persianPayload.code !== 200 || !persianPayload.data || !persianPayload.data.surahs) {
        console.error("❌ Invalid Persian API payload!");
        process.exit(1);
    }

    console.log("✅ Both Quran payloads downloaded successfully!");

    const surahsData = arabicPayload.data.surahs;
    const persianSurahs = persianPayload.data.surahs;

    const outputDir = path.join(__dirname, "..", "assets", "database");
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }

    const dbPath = path.join(outputDir, "quran.db");
    if (fs.existsSync(dbPath)) {
        fs.unlinkSync(dbPath);
        console.log(`🗑️ Removed old database at ${dbPath}`);
    }

    const db = new sqlite3.Database(dbPath);

    db.serialize(() => {
        console.log("🔨 Creating Database Tables & Indexes...");

        db.run("PRAGMA journal_mode = WAL;");

        // 1. Table: surahs
        db.run(`
            CREATE TABLE IF NOT EXISTS surahs (
                id INTEGER PRIMARY KEY,
                name TEXT NOT NULL,
                english_name TEXT NOT NULL,
                english_name_translation TEXT NOT NULL,
                number_of_ayahs INTEGER NOT NULL,
                revelation_type TEXT NOT NULL,
                start_page INTEGER NOT NULL,
                start_juz INTEGER NOT NULL
            );
        `);

        // 2. Table: ayahs
        db.run(`
            CREATE TABLE IF NOT EXISTS ayahs (
                id INTEGER PRIMARY KEY,
                surah_number INTEGER NOT NULL,
                number_in_surah INTEGER NOT NULL,
                text_uthmani TEXT NOT NULL,
                page INTEGER NOT NULL,
                juz INTEGER NOT NULL,
                hizb_quarter INTEGER NOT NULL,
                sajdah INTEGER DEFAULT 0,
                FOREIGN KEY (surah_number) REFERENCES surahs (id) ON DELETE CASCADE
            );
        `);

        db.run("CREATE INDEX IF NOT EXISTS idx_ayahs_surah ON ayahs(surah_number);");
        db.run("CREATE INDEX IF NOT EXISTS idx_ayahs_page ON ayahs(page);");
        db.run("CREATE INDEX IF NOT EXISTS idx_ayahs_juz ON ayahs(juz);");

        // 3. Table: translation_editions
        db.run(`
            CREATE TABLE IF NOT EXISTS translation_editions (
                identifier TEXT PRIMARY KEY,
                name TEXT NOT NULL,
                language TEXT NOT NULL,
                author TEXT,
                is_downloaded INTEGER DEFAULT 1
            );
        `);

        // 4. Table: translations
        db.run(`
            CREATE TABLE IF NOT EXISTS translations (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                translation_id TEXT NOT NULL,
                ayah_id INTEGER NOT NULL,
                text TEXT NOT NULL,
                FOREIGN KEY (translation_id) REFERENCES translation_editions (identifier) ON DELETE CASCADE,
                FOREIGN KEY (ayah_id) REFERENCES ayahs (id) ON DELETE CASCADE
            );
        `);

        db.run("CREATE INDEX IF NOT EXISTS idx_translations_lookup ON translations(translation_id, ayah_id);");

        // 5. Table: bookmarks
        db.run(`
            CREATE TABLE IF NOT EXISTS bookmarks (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                surah_number INTEGER NOT NULL,
                ayah_number INTEGER NOT NULL,
                page_number INTEGER NOT NULL,
                type TEXT NOT NULL,
                title TEXT,
                created_at INTEGER NOT NULL
            );
        `);

        // 6. Table: recitation_styles
        db.run(`
            CREATE TABLE IF NOT EXISTS recitation_styles (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                code TEXT NOT NULL UNIQUE,
                name TEXT NOT NULL
            );
        `);

        // 7. Table: reciters
        db.run(`
            CREATE TABLE IF NOT EXISTS reciters (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                identifier TEXT NOT NULL UNIQUE,
                name TEXT NOT NULL,
                english_name TEXT NOT NULL,
                arabic_name TEXT,
                subfolder TEXT NOT NULL,
                bitrate TEXT DEFAULT '128kbps',
                style_id INTEGER NOT NULL,
                FOREIGN KEY (style_id) REFERENCES recitation_styles (id) ON DELETE CASCADE
            );
        `);

        // Insert Translation Edition Metadata
        db.run(`
            INSERT INTO translation_editions (identifier, name, language, author, is_downloaded)
            VALUES ('fa.ansarian', 'استاد انصاریان', 'fa', 'حسین انصاریان', 1);
        `);

        // Insert Recitation Styles
        db.run(`
            INSERT INTO recitation_styles (id, code, name) VALUES
            (1, 'murattal', 'ترتیل'),
            (2, 'mujawwad', 'مجود'),
            (3, 'moallim', 'آموزشی (معلم)'),
            (4, 'translation', 'ترجمه صوتی');
        `);

        // Insert Verified Reciters Data
        db.run(`
            INSERT INTO reciters (identifier, name, english_name, arabic_name, subfolder, bitrate, style_id) VALUES
            ('parhizgar_48kbps', 'استاد شهریار پرهیزگار', 'Parhizgar', 'شهریار پرهیزگار', 'Parhizgar_48kbps', '48kbps', 1),
            ('alafasy_128kbps', 'استاد مشاری العفاسی', 'Mishary Alafasy', 'مشاري العفاسي', 'Alafasy_128kbps', '128kbps', 1),
            ('abdulsamad_murattal_192kbps', 'استاد عبدالباسط (ترتیل)', 'Abdul Basit (Murattal)', 'عبدالباسط عبدالصمد', 'Abdul_Basit_Murattal_192kbps', '192kbps', 1),
            ('abdulsamad_mujawwad_128kbps', 'استاد عبدالباسط (مجود)', 'Abdul Basit (Mujawwad)', 'عبدالباسط عبدالصمد', 'Abdul_Basit_Mujawwad_128kbps', '128kbps', 2),
            ('minshawy_murattal_128kbps', 'استاد محمد صدیق منشاوی (ترتیل)', 'Minshawy (Murattal)', 'محمد صديق المنشاوي', 'Minshawy_Murattal_128kbps', '128kbps', 1),
            ('minshawy_mujawwad_192kbps', 'استاد محمد صدیق منشاوی (مجود)', 'Minshawy (Mujawwad)', 'محمد صديق المنشاوي', 'Minshawy_Mujawwad_192kbps', '192kbps', 2),
            ('husary_128kbps', 'استاد محمود خلیل الحصری', 'Husary', 'محمود خليل الحصري', 'Husary_128kbps', '128kbps', 1),
            ('sudais_192kbps', 'استاد عبدالرحمن السدیس', 'Sudais', 'عبدالرحمن السديس', 'Abdurrahmaan_As-Sudais_192kbps', '192kbps', 1),
            ('ghamadi_40kbps', 'استاد سعد الغامدی', 'Ghamadi', 'سعد الغامدي', 'Ghamadi_40kbps', '40kbps', 1),
            ('shaatree_128kbps', 'استاد ابو بکر الشاطری', 'Abu Bakr Ash-Shaatree', 'أبو بكر الشاطري', 'Abu_Bakr_Ash-Shaatree_128kbps', '128kbps', 1),
            ('rifai_192kbps', 'استاد هانی الرفاعی', 'Hani Rifai', 'هاني الرفاعي', 'Hani_Rifai_192kbps', '192kbps', 1),
            ('hudhaify_128kbps', 'استاد علی بن عبدالرحمن الحذیفی', 'Hudhaify', 'علي بن عبدالرحمن الحذيفي', 'Hudhaify_128kbps', '128kbps', 1),
            ('shuraym_128kbps', 'استاد سعود الشریم', 'Saood ash-Shuraym', 'سعود الشريم', 'Saood_ash-Shuraym_128kbps', '128kbps', 1),
            ('maher_64kbps', 'استاد ماهر المعیقلی', 'Maher Al Muaiqly', 'ماهر المعيقلي', 'Maher_AlMuaiqly_64kbps', '64kbps', 1);
        `);

        console.log("📦 Populating Surahs, Ayahs, and Translations...");

        const insertSurahStmt = db.prepare(`
            INSERT INTO surahs (id, name, english_name, english_name_translation, number_of_ayahs, revelation_type, start_page, start_juz)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?);
        `);

        const insertAyahStmt = db.prepare(`
            INSERT INTO ayahs (id, surah_number, number_in_surah, text_uthmani, page, juz, hizb_quarter, sajdah)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?);
        `);

        const insertTranslationStmt = db.prepare(`
            INSERT INTO translations (translation_id, ayah_id, text)
            VALUES ('fa.ansarian', ?, ?);
        `);

        for (let i = 0; i < surahsData.length; i++) {
            const surah = surahsData[i];
            const pSurah = persianSurahs[i];

            const firstAyah = surah.ayahs[0];
            const startPage = firstAyah.page;
            const startJuz = firstAyah.juz;
            const numberOfAyahs = surah.ayahs.length;

            insertSurahStmt.run(
                surah.number,
                surah.name,
                surah.englishName,
                surah.englishNameTranslation,
                numberOfAyahs,
                surah.revelationType,
                startPage,
                startJuz
            );

            for (let j = 0; j < surah.ayahs.length; j++) {
                const ayah = surah.ayahs[j];
                const pAyah = pSurah.ayahs[j];

                let sajdaVal = 0;
                const sajdaRaw = ayah.sajda;
                if (typeof sajdaRaw === 'object' && sajdaRaw !== null) {
                    sajdaVal = sajdaRaw.obligatory ? 2 : 1;
                } else if (typeof sajdaRaw === 'boolean' && sajdaRaw) {
                    sajdaVal = 1;
                }

                insertAyahStmt.run(
                    ayah.number,
                    surah.number,
                    ayah.numberInSurah,
                    ayah.text,
                    ayah.page,
                    ayah.juz,
                    ayah.hizbQuarter,
                    sajdaVal
                );

                insertTranslationStmt.run(
                    ayah.number,
                    pAyah.text
                );
            }
        }

        insertSurahStmt.finalize();
        insertAyahStmt.finalize();
        insertTranslationStmt.finalize();

        db.get("SELECT COUNT(*) AS count FROM surahs", (err, row) => {
            console.log(`✨ Surahs Table Count: ${row.count}`);
        });

        db.get("SELECT COUNT(*) AS count FROM ayahs", (err, row) => {
            console.log(`✨ Ayahs Table Count: ${row.count}`);
        });

        db.get("SELECT COUNT(*) AS count FROM translations", (err, row) => {
            console.log(`✨ Translations Table Count: ${row.count}`);
        });

        db.get("SELECT COUNT(*) AS count FROM recitation_styles", (err, row) => {
            console.log(`✨ Recitation Styles Table Count: ${row.count}`);
        });

        db.get("SELECT COUNT(*) AS count FROM reciters", (err, row) => {
            console.log(`✨ Reciters Table Count: ${row.count}`);
        });
    });

    db.close((err) => {
        if (err) {
            console.error("❌ Error closing DB:", err);
            process.exit(1);
        }
        const stats = fs.statSync(dbPath);
        const sizeMb = (stats.size / (1024 * 1024)).toFixed(2);
        console.log("\n🎉 SUCCESS! quran.db created successfully!");
        console.log(`📍 Output Path: ${dbPath}`);
        console.log(`📊 File Size: ${sizeMb} MB`);
    });
}

main().catch(err => {
    console.error("❌ Unhandled Error:", err);
    process.exit(1);
});

File: scripts/generate_quran_db.py
import sqlite3
import urllib.request
import json
import os
import sys

def build_quran_database():
    print("🚀 Starting Quran Database Generation...")
    
    # 1. API Endpoints
    url = "https://api.alquran.cloud/v1/quran/quran-uthmani,fa.ansarian"
    
    print(f"📥 Fetching full Quran payload from: {url}")
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    
    with urllib.request.urlopen(req) as response:
        if response.status != 200:
            print(f"❌ Failed to fetch data: HTTP {response.status}")
            sys.exit(1)
        raw_data = response.read().decode('utf-8')
        payload = json.loads(raw_data)
        
    if payload.get("code") != 200 or "data" not in payload:
        print("❌ Invalid API response payload!")
        sys.exit(1)
        
    print("✅ Quran payload downloaded successfully!")
    
    # Extract Uthmani Arabic edition and Persian Ansarian edition
    data = payload["data"]
    arabic_edition = data[0]     # quran-uthmani
    persian_edition = data[1]    # fa.ansarian
    
    # Ensure assets/database directory exists
    output_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "assets", "database")
    os.makedirs(output_dir, exist_ok=True)
    db_path = os.path.join(output_dir, "quran.db")
    
    # Delete old database if exists
    if os.path.exists(db_path):
        os.remove(db_path)
        print(f"🗑️ Removed existing database at {db_path}")

    # 2. Connect to SQLite
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    
    # Enable WAL mode for performance
    cursor.execute("PRAGMA journal_mode = WAL;")
    
    # 3. Create Tables
    print("🔨 Creating Database Tables & Indexes...")
    
    cursor.execute("""
    CREATE TABLE IF NOT EXISTS surahs (
        id INTEGER PRIMARY KEY,
        name TEXT NOT NULL,
        english_name TEXT NOT NULL,
        english_name_translation TEXT NOT NULL,
        number_of_ayahs INTEGER NOT NULL,
        revelation_type TEXT NOT NULL,
        start_page INTEGER NOT NULL,
        start_juz INTEGER NOT NULL
    );
    """)

    cursor.execute("""
    CREATE TABLE IF NOT EXISTS ayahs (
        id INTEGER PRIMARY KEY,
        surah_number INTEGER NOT NULL,
        number_in_surah INTEGER NOT NULL,
        text_uthmani TEXT NOT NULL,
        page INTEGER NOT NULL,
        juz INTEGER NOT NULL,
        hizb_quarter INTEGER NOT NULL,
        sajdah INTEGER DEFAULT 0,
        FOREIGN KEY (surah_number) REFERENCES surahs (id) ON DELETE CASCADE
    );
    """)
    
    cursor.execute("CREATE INDEX IF NOT EXISTS idx_ayahs_surah ON ayahs(surah_number);")
    cursor.execute("CREATE INDEX IF NOT EXISTS idx_ayahs_page ON ayahs(page);")
    cursor.execute("CREATE INDEX IF NOT EXISTS idx_ayahs_juz ON ayahs(juz);")

    cursor.execute("""
    CREATE TABLE IF NOT EXISTS translation_editions (
        identifier TEXT PRIMARY KEY,
        name TEXT NOT NULL,
        language TEXT NOT NULL,
        author TEXT,
        is_downloaded INTEGER DEFAULT 1
    );
    """)

    cursor.execute("""
    CREATE TABLE IF NOT EXISTS translations (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        translation_id TEXT NOT NULL,
        ayah_id INTEGER NOT NULL,
        text TEXT NOT NULL,
        FOREIGN KEY (translation_id) REFERENCES translation_editions (identifier) ON DELETE CASCADE,
        FOREIGN KEY (ayah_id) REFERENCES ayahs (id) ON DELETE CASCADE
    );
    """)

    cursor.execute("CREATE INDEX IF NOT EXISTS idx_translations_lookup ON translations(translation_id, ayah_id);")

    cursor.execute("""
    CREATE TABLE IF NOT EXISTS bookmarks (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        surah_number INTEGER NOT NULL,
        ayah_number INTEGER NOT NULL,
        page_number INTEGER NOT NULL,
        type TEXT NOT NULL,
        title TEXT,
        created_at INTEGER NOT NULL
    );
    """)

    # 4. Insert Translation Edition Metadata
    cursor.execute("""
    INSERT INTO translation_editions (identifier, name, language, author, is_downloaded)
    VALUES ('fa.ansarian', 'استاد انصاریان', 'fa', 'حسین انصاریان', 1);
    """)

    # 5. Populate Surahs, Ayahs, and Translations
    print("📦 Populating Surahs, Ayahs, and Translations...")
    
    surahs_data = arabic_edition["surahs"]
    persian_surahs = persian_edition["surahs"]
    
    total_ayah_count = 0
    total_translation_count = 0

    for i in range(len(surahs_data)):
        surah = surahs_data[i]
        p_surah = persian_surahs[i]
        
        surah_num = surah["number"]
        surah_name = surah["name"]
        english_name = surah["englishName"]
        english_name_translation = surah["englishNameTranslation"]
        number_of_ayahs = surah["numberOfAyahs"]
        revelation_type = surah["revelationType"]
        
        # Determine start page and start juz from the 1st ayah
        first_ayah = surah["ayahs"][0]
        start_page = first_ayah["page"]
        start_juz = first_ayah["juz"]
        
        cursor.execute("""
        INSERT INTO surahs (id, name, english_name, english_name_translation, number_of_ayahs, revelation_type, start_page, start_juz)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?);
        """, (surah_num, surah_name, english_name, english_name_translation, number_of_ayahs, revelation_type, start_page, start_juz))
        
        # Populate Ayahs & Translations for this surah
        for j in range(len(surah["ayahs"])):
            ayah = surah["ayahs"][j]
            p_ayah = p_surah["ayahs"][j]
            
            global_ayah_id = ayah["number"]
            number_in_surah = ayah["numberInSurah"]
            text_uthmani = ayah["text"]
            page = ayah["page"]
            juz = ayah["juz"]
            hizb_quarter = ayah["hizbQuarter"]
            
            # Sajdah calculation: False -> 0, True -> 1 (Recommended) or 2 (Obligatory)
            sajda_val = 0
            sajda_raw = ayah.get("sajda", False)
            if isinstance(sajda_raw, dict):
                sajda_val = 2 if sajda_raw.get("obligatory", False) else 1
            elif isinstance(sajda_raw, bool) and sajda_raw:
                sajda_val = 1
                
            cursor.execute("""
            INSERT INTO ayahs (id, surah_number, number_in_surah, text_uthmani, page, juz, hizb_quarter, sajdah)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?);
            """, (global_ayah_id, surah_num, number_in_surah, text_uthmani, page, juz, hizb_quarter, sajda_val))
            
            total_ayah_count += 1
            
            # Insert Translation Text
            translation_text = p_ayah["text"]
            cursor.execute("""
            INSERT INTO translations (translation_id, ayah_id, text)
            VALUES ('fa.ansarian', ?, ?);
            """, (global_ayah_id, translation_text))
            
            total_translation_count += 1

    conn.commit()
    conn.close()
    
    db_size_mb = os.path.getsize(db_path) / (1024 * 1024)
    print("\n🎉 SUCCESS! Database built successfully!")
    print(f"📍 Database File Path: {db_path}")
    print(f"📊 Database Size: {db_size_mb:.2f} MB")
    print(f"✨ Total Surahs Inserted: {len(surahs_data)}")
    print(f"✨ Total Ayahs Inserted: {total_ayah_count}")
    print(f"✨ Total Translations Inserted: {total_translation_count}")

if __name__ == "__main__":
    build_quran_database()
